### PR TITLE
Add j.l.Character.to{Lower,Upper}Case(codePoint: Int)/toTitleCase

### PR DESCRIPTION
**Pull request by er1c**
Addresses: https://github.com/scala-js/scala-js/issues/4206 & https://github.com/scala-js/scala-js/issues/4208

Also fixes minor edge case in `Character.toUpperCase(ch: Char)`

TODO:
- [x] Add a test character that expands into a multiple-codepoint via Strinig.toUpperCase/toLowerCase 

> Then I'm not sure what we should do with results that span more than one code point. It's not necessarily just `length > 2`; it's also possible that this happens with `length == 2` if the two Chars are not a surrogate pair. We need to carefully check the spec for that.

### Javadoc

	public static int toUpperCase(int codePoint)
    
	Converts the character (Unicode code point) argument to uppercase using case mapping information from the UnicodeData file.
	Note that Character.isUpperCase(Character.toUpperCase(codePoint)) does not always return true for some ranges of characters, particularly those that are symbols or ideographs.
	
	In general, String.toUpperCase() should be used to map characters to uppercase. String case mapping methods have several benefits over Character case mapping methods. String case mapping methods can perform locale-sensitive mappings, context-sensitive mappings, and 1:M character mappings, whereas the Character case mapping methods cannot.
	
	Parameters:
	codePoint - the character (Unicode code point) to be converted.
	Returns:
	the uppercase equivalent of the character, if any; otherwise, the character itself.
	Since:
	1.5
	See Also:
	isUpperCase(int), String.toUpperCase()

### Implementation notes

`Character.isSurrogatePair` can be used to verify length == 2, and only return the codePoint if they are a pair, otherwise the original codePoint is suppose to be returned

### CODE REVIEW

**Comment by er1c** [comment]
Head scratching: https://scastie.scala-lang.org/rrBuaGmkRry8cllxYWwDCA

codePoint - Int, codePoint - hex
```
223 - 0x00df - strUpperCase: SS, chrUpperCase: ß
329 - 0x0149 - strUpperCase: ʼN, chrUpperCase: ŉ
496 - 0x01f0 - strUpperCase: J̌, chrUpperCase: ǰ
912 - 0x0390 - strUpperCase: Ϊ́, chrUpperCase: ΐ
944 - 0x03b0 - strUpperCase: Ϋ́, chrUpperCase: ΰ
1415 - 0x0587 - strUpperCase: ԵՒ, chrUpperCase: և
7830 - 0x1e96 - strUpperCase: H̱, chrUpperCase: ẖ
7831 - 0x1e97 - strUpperCase: T̈, chrUpperCase: ẗ
7832 - 0x1e98 - strUpperCase: W̊, chrUpperCase: ẘ
7833 - 0x1e99 - strUpperCase: Y̊, chrUpperCase: ẙ
```

I suspect there just aren't tests in place today to show a current failing Character.toUpperCase(0x00df.toChar) against the JVM

**Comment by er1c** [comment]
Added a new test that fails against master, but works against this pr:

```
  @Test def toUpperCase_Char_MultiCodePoints(): Unit = {
    assertEquals('ΐ', Character.toUpperCase('ΐ'))
    assertEquals('ΰ', Character.toUpperCase('ΰ'))
    assertEquals('ὒ', Character.toUpperCase('ὒ'))
    assertEquals('ὔ', Character.toUpperCase('ὔ'))
    assertEquals('ὖ', Character.toUpperCase('ὖ'))
    assertEquals('ᾷ', Character.toUpperCase('ᾷ'))
    assertEquals('ῇ', Character.toUpperCase('ῇ'))
    assertEquals('ῒ', Character.toUpperCase('ῒ'))
    assertEquals('ΐ', Character.toUpperCase('ΐ'))
    assertEquals('ῗ', Character.toUpperCase('ῗ'))
    assertEquals('ῢ', Character.toUpperCase('ῢ'))
    assertEquals('ΰ', Character.toUpperCase('ΰ'))
    assertEquals('ῧ', Character.toUpperCase('ῧ'))
    assertEquals('ῷ', Character.toUpperCase('ῷ'))
    assertEquals('ﬃ', Character.toUpperCase('ﬃ'))
    assertEquals('ﬄ', Character.toUpperCase('ﬄ'))
  }
```

**Comment by er1c** [comment]
I added `toTitleCase` which is based upon this PR as well

File: javalanglib/src/main/scala/java/lang/Character.scala [hunk 1]
```diff
@@ -462,9 +462,90 @@ object Character {
...
+  val titleCaseCP: Int = Character.toTitleCase(cp)
+  val upperCaseCP: Int = Character.toUpperCase(cp)
+
+  if (titleCaseCP != upperCaseCP) {
+    println(s"    case ${format(cp)} => ${format(titleCaseCP)}")
+  }
+}
+*/
+  def toTitleCase(codePoint: scala.Int): scala.Int = {
+    codePoint match {
```
Comment (er1c) [inline javalanglib/src/main/scala/java/lang/Character.scala:532]: @sjrd does adding a `@switch` here matter, and a better question is how can I verify this myself?

File: javalanglib/src/main/scala/java/lang/Character.scala [hunk 2]
```diff
@@ -462,8 +462,54 @@ object Character {
...
+    val upperChars = _String.fromCodePoint(codePoint).toUpperCase
+
+    if (upperChars.length > 2) {
+      codePoint
+    } else if (upperChars.length == 2) {
+      val high = upperChars.charAt(0)
+      val low = upperChars.charAt(1)
+
+      if (isSurrogatePair(high, low))
+        toCodePoint(upperChars.charAt(0), upperChars.charAt(1))
```
Comment (sjrd) [inline javalanglib/src/main/scala/java/lang/Character.scala:488]: Reuse `high` and `low`?
```suggestion
        toCodePoint(high, low)
```

File: javalanglib/src/main/scala/java/lang/Character.scala [hunk 3]
```diff
@@ -462,8 +462,54 @@ object Character {
...
+  }
+
+  def toLowerCase(ch: scala.Char): scala.Char = {
+    val lower = toLowerCase(ch.toInt)
+    if (lower > scala.Char.MaxValue) ch
+    else lower.toChar
+  }
+
+  def toUpperCase(codePoint: scala.Int): scala.Int = {
+    val upperChars = _String.fromCodePoint(codePoint).toUpperCase
```
Comment (sjrd) [inline javalanglib/src/main/scala/java/lang/Character.scala:479]:
```suggestion
    val upperChars = _String.fromCodePoint(codePoint).toUpperCase()
```

File: javalanglib/src/main/scala/java/lang/Character.scala [hunk 4]
```diff
@@ -462,8 +462,54 @@ object Character {
...
+        toCodePoint(upperChars.charAt(0), upperChars.charAt(1))
+      else
+        codePoint
+    } else {
+      upperChars.charAt(0).toInt
+    }
+  }
+
+  def toLowerCase(codePoint: scala.Int): scala.Int = {
+    val lowerChars = _String.fromCodePoint(codePoint).toLowerCase
```
Comment (sjrd) [inline javalanglib/src/main/scala/java/lang/Character.scala:497]:
```suggestion
    val lowerChars = _String.fromCodePoint(codePoint).toLowerCase()
```

File: javalanglib/src/main/scala/java/lang/Character.scala [hunk 5]
```diff
@@ -462,8 +462,54 @@ object Character {
...
+
+    if (lowerChars.length > 2) {
+      codePoint
+    } else if (lowerChars.length == 2) {
+      val high = lowerChars.charAt(0)
+      val low = lowerChars.charAt(1)
+
+      if (isSurrogatePair(high, low))
+        toCodePoint(lowerChars.charAt(0), lowerChars.charAt(1))
```
Comment (sjrd) [inline javalanglib/src/main/scala/java/lang/Character.scala:506]:
```suggestion
        toCodePoint(high, low)
```

File: javalanglib/src/main/scala/java/lang/Character.scala [hunk 6]
```diff
@@ -462,8 +462,54 @@ object Character {
...
+    if (lower > scala.Char.MaxValue) ch
+    else lower.toChar
+  }
+
+  def toUpperCase(codePoint: scala.Int): scala.Int = {
+    val upperChars = _String.fromCodePoint(codePoint).toUpperCase
+
+    if (upperChars.length > 2) {
+      codePoint
+    } else if (upperChars.length == 2) {
```
Comment (sjrd) [inline javalanglib/src/main/scala/java/lang/Character.scala:483]: Instead of this chain of if/else's, consider using a match:
```scala
upperChars.length match {
  case 1 =>
    upperChars.charAt(0).toInt
  case 2 =>
    ...
  case _ =>
    codePoint
}
```

File: test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/lang/CharacterTest.scala [hunk 7]
```diff
@@ -383,6 +383,159 @@ class CharacterTest {
...
+  @Test def toLowerCase_Char(): Unit = {
+    // Low
+    assertEquals('\n', Character.toLowerCase('\n'))
+    // Ascii
+    assertEquals('0', Character.toLowerCase('0'))
+    assertEquals('a', Character.toLowerCase('a'))
+    assertEquals('a', Character.toLowerCase('A'))
+    assertEquals('f', Character.toLowerCase('F'))
+    assertEquals('z', Character.toLowerCase('Z'))
+    // compat/ucd
```
Comment (sjrd) [inline test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/lang/CharacterTest.scala:395]: These comments don't mean anything to me. What do they refer to?

Consider using comments that actually explain why the tests are interesting. I could see a few reasons:

* ASCII repertoire, not a letter (no change)
* ASCII repertoire, lowercase letter (no change)
* ASCII repertoire, uppercase letter
* Somewhere else in the BMP, not a letter (no change)
* Somewhere else in the BMP, lowercase letter (no change)
* Somewhere else in the BMP, titlecase letter
* Somewhere else in the BMP, uppercase letter
* Surrogate Chars (no change)

File: test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/lang/CharacterTest.scala [hunk 8]
```diff
@@ -383,6 +383,159 @@ class CharacterTest {
...
+    assertEquals('į', Character.toLowerCase('Į'))
+    // high
+    assertEquals('ａ', Character.toLowerCase('Ａ'))
+    assertEquals('ｚ', Character.toLowerCase('Ｚ'))
+    assertEquals('［', Character.toLowerCase('［'))
+  }
+
+  @Test def toLowerCase_Int(): Unit = {
+    // Low
+    assertEquals(10, Character.toLowerCase(10)) // '\n' => '\n'
```
Comment (sjrd) [inline test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/lang/CharacterTest.scala:412]: Please consistently use hexadecimal numbers for code points.

File: test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/lang/CharacterTest.scala [hunk 9]
```diff
@@ -383,6 +383,159 @@ class CharacterTest {
...
+    // Ascii
+    assertEquals(97, Character.toLowerCase(97)) // a => a
+    assertEquals(97, Character.toLowerCase(65)) // A => a
+    assertEquals(102, Character.toLowerCase(70)) // F => f
+    assertEquals(122, Character.toLowerCase(90)) // Z => z
+    // compat/ucd
```
Comment (sjrd) [inline test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/lang/CharacterTest.scala:418]: Same issue with the comments. For `toLowerCase(int)`, there are additional interesting categories:

* Outside the BMP, not a letter (no change)
* Outside the BMP, lowercase (no change)
* (Outside the BMP, titlecase -- not sure that exists)
* Outside the BMP, uppercase

File: test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/lang/CharacterTest.scala [hunk 10]
```diff
@@ -383,6 +383,159 @@ class CharacterTest {
...
+    assertEquals(258, Character.toUpperCase(259))
+    assertEquals(302, Character.toUpperCase(303))
+    // high
+    assertEquals(65313, Character.toUpperCase(65345))
+    assertEquals(65338, Character.toUpperCase(65370))
+    assertEquals(65339, Character.toUpperCase(65339))
+  }
+
+/*
+for(cp <- 0 until Character.MAX_CODE_POINT; if cp <= Character.MAX_VALUE) {
```
Comment (sjrd) [inline test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/lang/CharacterTest.scala:522]: Here we should iterate over Chars, not over code points. Chars are not code points (they are code *units*):
```suggestion
for (c <- 0.toChar until Character.MAX_VALUE) {
```

File: test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/lang/CharacterTest.scala [hunk 11]
```diff
@@ -383,6 +383,159 @@ class CharacterTest {
...
+  val strUpperCase: String = {
+    val s = new String(Array(cp), 0, 1)
+    s.toUpperCase()
+  }
+
+  if (strUpperCase.codePointCount(0, strUpperCase.length-1) > 1)
```
Comment (sjrd) [inline test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/lang/CharacterTest.scala:528]: Here we should also include anything that spans more than 1 Char (1 code unit), since they cannot be represented in the result either:
```suggestion
  if (strUpperCase.length > 1)
```

File: test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/lang/CharacterTest.scala [hunk 12]
```diff
@@ -383,6 +383,159 @@ class CharacterTest {
...
+    assertEquals('ῧ', Character.toUpperCase('ῧ'))
+    assertEquals('ῷ', Character.toUpperCase('ῷ'))
+    assertEquals('ﬃ', Character.toUpperCase('ﬃ'))
+    assertEquals('ﬄ', Character.toUpperCase('ﬄ'))
+  }
+
+/*
+def format(codePoint: Int): String = "0x%04x".format(codePoint)
+
+for(cp <- 0 until Character.MAX_CODE_POINT) {
```
Comment (sjrd) [inline test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/lang/CharacterTest.scala:554]:
```suggestion
for (cp <- 0 until Character.MAX_CODE_POINT) {
```

File: test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/lang/CharacterTest.scala [hunk 13]
```diff
@@ -383,6 +383,159 @@ class CharacterTest {
...
+  val strUpperCase: String = {
+    val s = new String(Array(cp), 0, 1)
+    s.toUpperCase()
+  }
```
Comment (sjrd) [inline test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/lang/CharacterTest.scala:558]: Consider making this shorter:
```suggestion
  val strUpperCase: String = new String(Array(cp), 0, 1).toUpperCase()
```

File: test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/lang/CharacterTest.scala [hunk 14]
```diff
@@ -383,6 +383,159 @@ class CharacterTest {
...
+
+  if (strUpperCase.codePointCount(0, strUpperCase.length-1) > 1)
```
Comment (sjrd) [inline test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/lang/CharacterTest.scala:560]: There must not be a `-1` here. The second argument to `codePointCount` is interpreted as an end *excluded*:
```suggestion
  if (strUpperCase.codePointCount(0, strUpperCase.length) > 1)
```
(this could actually change the output of the script)

File: javalanglib/src/main/scala/java/lang/Character.scala [hunk 15]
```diff
@@ -510,7 +511,42 @@ object Character {
...
+  def toTitleCase(ch: scala.Char): scala.Char = {
+    val title = toTitleCase(ch.toInt)
+    if (title > scala.Char.MaxValue) ch
+    else title.toChar
+  }
+
+/*
+def format(codePoint: Int): String = "0x%04x".format(codePoint)
+
+for(cp <- 0 until Character.MAX_CODE_POINT) {
```
Comment (sjrd) [inline javalanglib/src/main/scala/java/lang/Character.scala:523]:
```suggestion
for (cp <- 0 until Character.MAX_CODE_POINT) {
```

File: javalanglib/src/main/scala/java/lang/Character.scala [hunk 16]
```diff
@@ -510,7 +511,42 @@ object Character {
...
+      case 0x01c7 => 0x01c8
+      case 0x01c8 => 0x01c8
+      case 0x01c9 => 0x01c8
+      case 0x01ca => 0x01cb
+      case 0x01cb => 0x01cb
+      case 0x01cc => 0x01cb
+      case 0x01f1 => 0x01f2
+      case 0x01f2 => 0x01f2
+      case 0x01f3 => 0x01f2
+      case _ => toUpperCase(codePoint)
```
Comment (sjrd) [inline javalanglib/src/main/scala/java/lang/Character.scala:546]:
```suggestion
      case 0x01f3 => 0x01f2
      case _      => toUpperCase(codePoint)
```

File: test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/lang/CharacterTest.scala [hunk 17]
```diff
@@ -429,6 +429,50 @@ class CharacterTest {
...
+    if (cp <= Character.MAX_VALUE) {
+      assert(titleCaseCP <= Character.MAX_VALUE)
+      println(s"    assertEquals('${titleCaseCP.toChar}', Character.toTitleCase('${cp.toChar}'))")
+    }
+
+    println(s"    assertEquals(${format(titleCaseCP)}, Character.toTitleCase(${format(cp)}))")
+  }
+}
+*/
+  @Test def toTitleCase(): Unit  = {
```
Comment (sjrd) [inline test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/lang/CharacterTest.scala:449]: We should definitely also test cases where the result is the same as `toUpperCase`. Basically you can add all the tests written for `toUpperCase` (including the ones that get "aborted" because the result does not fit in 1 char/1 code point) but for `toTitleCase`.

**Review by sjrd — CHANGES_REQUESTED** [review]
The title of the first commit message is too long. We often use `{xy,zw}` to deal with this:

    Add j.l.Character.to{Lower,Upper}Case(codePoint: Int)

File: test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/lang/CharacterTest.scala (hunk 9 above)
Comment (er1c) [reply test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/lang/CharacterTest.scala:418]: Dumb question, what's BMP?

Comment (sjrd) [reply test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/lang/CharacterTest.scala:418]: [Basic Multilingual Plane](https://en.wikipedia.org/wiki/Plane_(Unicode)#Basic_Multilingual_Plane). It's the code points that fit in 1 UTF-16 code unit. That's opposed to the code points `>= 0x10000`, which require a surrogate pair.

File: test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/lang/CharacterTest.scala (hunk 17 above)
Comment (er1c) [reply test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/lang/CharacterTest.scala:449]: FYI, this logic got updated based upon additional testing (well the `toUpperCase(Char)` got fixed) but yes I still should add tests here,  I just didn't want you to overlook the implementation change to `def toTitleCase(ch: scala.Char): scala.Char`

File: test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/lang/CharacterTest.scala (hunk 7 above)
Comment (sjrd) [reply test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/lang/CharacterTest.scala:395]: Given the new implementations I suggest above, we should definitely have exhaustive tests for all the special characters in `toUpperCase` and `toTitleCase`.

We could also have 3 generic sanity tests for the overloads with Chars, of the form:
```scala
for (ch <- Character.MIN_VALUE to Character.MAX_VALUE)
  assertEquals(Character.toLowerCase(ch), Character.toLowerCase(ch.toInt).toChar)
```
(with the same for `toUpperCase` and `toTitleCase`). When we run those on the JVM, we make sure that the naive implementations we have for the Char overloads is indeed correct.

**Review by sjrd — CHANGES_REQUESTED** [review]

**Comment by er1c** [comment]
@sjrd PR updated, waiting on CI, thanks for coming up with a more elegant `toUpperCase` I've updated the tests as requested as well

File: test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/lang/CharacterTest.scala [hunk 18]
```diff
@@ -434,9 +434,136 @@ class CharacterTest {
...
+    assertEquals('ǲ', Character.toTitleCase('Ǳ'))
+    assertEquals('ǲ', Character.toTitleCase('ǲ'))
+    assertEquals('ǲ', Character.toTitleCase('ǳ'))
+  }
+
+/*
+def format(codePoint: Int): String = "0x%04x".format(codePoint)
+
+def myToUpperCase(cp: Int): Int = {
+  val upperString = new String(Array(cp),0,1).toUpperCase()
```
Comment (sjrd) [inline test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/lang/CharacterTest.scala:470]:
```suggestion
  val upperString = new String(Array(cp), 0, 1).toUpperCase()
```

File: test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/lang/CharacterTest.scala [hunk 19]
```diff
@@ -434,9 +434,136 @@ class CharacterTest {
...
+    upperString.charAt(0)
+  else
+    cp
+}
+
+for (cp <- 0 to Character.MAX_CODE_POINT) {
+  if (myToUpperCase(cp) != Character.toTitleCase(cp)) {
+    val cpStr: String  = new String(Array(cp),0,1)
+    val titleCP: Int = Character.toTitleCase(cp)
+    val titleStr: String = new String(Array(titleCP),0,1)
```
Comment (sjrd) [inline test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/lang/CharacterTest.scala:481]:
```suggestion
    val cpStr: String  = new String(Array(cp), 0, 1)
    val titleCP: Int = Character.toTitleCase(cp)
    val titleStr: String = new String(Array(titleCP), 0, 1)
```

File: test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/lang/CharacterTest.scala [hunk 20]
```diff
@@ -434,9 +434,136 @@ class CharacterTest {
...
+    assertEquals('ǲ', Character.toTitleCase('ǳ'))
+  }
+
+/*
+def format(codePoint: Int): String = "0x%04x".format(codePoint)
+
+def myToUpperCase(cp: Int): Int = {
+  val upperString = new String(Array(cp),0,1).toUpperCase()
+  if (upperString.length() == 1)
+    upperString.charAt(0)
```
Comment (sjrd) [inline test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/lang/CharacterTest.scala:472]: Should this be expressed in terms of code points?
```suggestion
  if (upperString.codePointCount(0, upperString.length()) == 1)
    upperString.charPointAt(0)
```
In fact, I don't think I understand what codepoints are selected by this script and why?

File: test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/lang/CharacterTest.scala [hunk 21]
```diff
@@ -434,9 +434,136 @@ class CharacterTest {
...
+/*
+for (ch <- 0.toChar until Character.MAX_VALUE) {
+  val titleCase: Char = Character.toTitleCase(ch)
+  val upperCase: Char = Character.toUpperCase(ch)
+
+  if (titleCase != upperCase)
+    println(s"    assertEquals('$titleCase', Character.toTitleCase('$ch'))")
+}
+*/
+  @Test def toTitleCase_Char_UpperCase_Differences(): Unit = {
```
Comment (sjrd) [inline test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/lang/CharacterTest.scala:451]: I now realized that, since we have the exhaustive loop in `toTitleCase_compare_char_and_codepoint()`, we don't need to test anything else about `toTitleCase(Char)`. All the special cases should be tested in `toTitleCase(Int)` anyway, so this is redundant.

The rationale is: `toTitleCase_compare_char_and_codepoint()` can basically assume that `toTitleCase(Int)` is correct. If that one is correct, and we exhaustively test `toTitleCase(Char)` against `toTitleCase(Int)`, then `toTitleCase(Char)` is correct.

(and the test on the JVM will ensure that the *way* we test `totTitleCase(Char)` against `toTitleCase(Int)` is correct itself)

File: javalanglib/src/main/scala/java/lang/Character.scala [hunk 22]
```diff
@@ -462,9 +463,206 @@ object Character {
...
+  def toLowerCase(codePoint: scala.Int): scala.Int = {
+    val lowerChars = _String.fromCodePoint(codePoint).toLowerCase()
+
+    lowerChars.length match {
+      case 1 =>
+        lowerChars.charAt(0).toInt
+      case 2 =>
+        val high = lowerChars.charAt(0)
+        val low = lowerChars.charAt(1)
+
```
[the remark this answers is not in the document]
Comment (sjrd) [reply javalanglib/src/main/scala/java/lang/Character.scala:513]: This is still pending.

File: javalanglib/src/main/scala/java/lang/Character.scala [hunk 23]
```diff
@@ -462,8 +462,63 @@ object Character {
...
+/*
+for (ch <- 0.toChar until Character.MAX_VALUE) {
+  val strLowerCase: String = ch.toString.toLowerCase()
+  val chrLowerCase: Char = Character.toLowerCase(ch)
+
+  if (strLowerCase.charAt(0) != chrLowerCase)
+    println(s"    case '$ch' => '$chrLowerCase'")
+}
+// No Results
+*/
```
Comment (sjrd) [inline javalanglib/src/main/scala/java/lang/Character.scala:501]: This comment is not necessary anymore. We've established that it is in fact true for the 3 methods: `toUpperCase`, `toLowerCase` and `toTitleCase`. Writing this comment here but not for the other two is therefore misleading. We could also write that kind of comment on the other two instead, but that would be completely redundant with the exhaustive tests, so that's not useful IMO.

File: test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/lang/CharacterTest.scala [hunk 24]
```diff
@@ -434,9 +434,136 @@ class CharacterTest {
     assertEquals(0xff3b, Character.toLowerCase(0xff3b)) // ［ => ［
   }
 
-  @Test def toUpperCase_compare_char_and_codepoint(): Unit = {
+  @Test def toTitleCase_compare_char_and_codepoint(): Unit = {
     for (ch <- Character.MIN_VALUE to Character.MAX_VALUE)
-      assertEquals(Character.toUpperCase(ch), Character.toUpperCase(ch.toInt).toChar)
+      assertEquals(Character.toTitleCase(ch), Character.toTitleCase(ch.toInt).toChar)
+  }
```
Comment (sjrd) [inline test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/lang/CharacterTest.scala]: There seems to be a rebasing problem here. The second commit *removes* the test for `toUpperCase_compare_char_and_codepoint()`. It should be kept.

File: test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/lang/CharacterTest.scala [hunk 25]
```diff
@@ -383,6 +383,559 @@ class CharacterTest {
     assertFalse(Character.isDigit('\uFBFC'))
   }
 
+  @Test def toLowerCase_compare_char_and_codepoint(): Unit = {
+    for (ch <- Character.MIN_VALUE to Character.MAX_VALUE)
+      assertEquals(Character.toLowerCase(ch), Character.toLowerCase(ch.toInt).toChar)
+  }
+
+  @Test def toLowerCase_Char(): Unit = {
```
Comment (sjrd) [inline test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/lang/CharacterTest.scala:391]: Similarly to what I said for `toTitleCase`, this test is basically redundant with `toLowerCase_Int()`, now, so it can be removed.

File: test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/lang/CharacterTest.scala [hunk 26]
```diff
@@ -383,6 +383,432 @@ class CharacterTest {
...
+    assertEquals(0xff5a, Character.toLowerCase(0xff3a)) // Ｚ => ｚ
+    assertEquals(0xff3b, Character.toLowerCase(0xff3b)) // ［ => ［
+  }
+
+  @Test def toUpperCase_compare_char_and_codepoint(): Unit = {
+    for (ch <- Character.MIN_VALUE to Character.MAX_VALUE)
+      assertEquals(Character.toUpperCase(ch), Character.toUpperCase(ch.toInt).toChar)
+  }
+
+  @Test def toUpperCase_Char(): Unit = {
```
Comment (sjrd) [inline test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/lang/CharacterTest.scala:569]: And same thing here: this is redundant and can be removed.

File: test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/lang/CharacterTest.scala (hunk 9 above)
Comment (sjrd) [reply test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/lang/CharacterTest.scala:423]: This is still pending. I'd still like to see tests broken down by categories that actually make sense. Combining what I wrote at https://github.com/scala-js/scala-js/pull/4207/commits/b9438ca40bccb2035cf31af613c459b559c8ded5#r494129429 and what is three comments above, a full list would be:

* ASCII repertoire, not a letter (no change)
* ASCII repertoire, lowercase letter (no change)
* ASCII repertoire, uppercase letter
* Somewhere else in the BMP, not a letter (no change)
* Somewhere else in the BMP, lowercase letter (no change)
* Somewhere else in the BMP, titlecase letter
* Somewhere else in the BMP, uppercase letter
* A high surrogate and a low surrogate (no change)
* Outside the BMP, not a letter (no change)
* Outside the BMP, lowercase (no change)
* (Outside the BMP, titlecase -- not sure that exists)
* Outside the BMP, uppercase

There can be just one of each of those.

Then the same would go for `toUpperCase(Int)` and `toTitleCase(Int)` (the exact same list of inputs can be reused, with different outputs). With that and the existing `toUpperCase_Int_special` and `toTitleCase_Int_special`, we'll have a good coverage.

File: test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/lang/CharacterTest.scala [hunk 27]
```diff
@@ -383,6 +383,432 @@ class CharacterTest {
...
+  @Test def toUpperCase_Int_special(): Unit = {
+    assertEquals(0x1f88, Character.toUpperCase(0x1f80)) // ᾀ => ᾈ
+    assertEquals(0x1f89, Character.toUpperCase(0x1f81)) // ᾁ => ᾉ
+    assertEquals(0x1f8a, Character.toUpperCase(0x1f82)) // ᾂ => ᾊ
+    assertEquals(0x1f8b, Character.toUpperCase(0x1f83)) // ᾃ => ᾋ
+    assertEquals(0x1f8c, Character.toUpperCase(0x1f84)) // ᾄ => ᾌ
+    assertEquals(0x1f8d, Character.toUpperCase(0x1f85)) // ᾅ => ᾍ
+    assertEquals(0x1f8e, Character.toUpperCase(0x1f86)) // ᾆ => ᾎ
+    assertEquals(0x1f8f, Character.toUpperCase(0x1f87)) // ᾇ => ᾏ
+    assertEquals(0x1f98, Character.toUpperCase(0x1f90)) // ᾐ => ᾘ
```
Comment (sjrd) [inline test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/lang/CharacterTest.scala:644]: We should also test `0x1f88..0x1f8f` as *inputs* here, since our algorithm captures those as special case inputs. We need to make sure that our trick with the `| 0x0008` does the right thing for the inputs in that range that are already uppercase.

(same with the other 2 ranges of uppercases)

**Review by sjrd — CHANGES_REQUESTED** [review]
OK, we're definitely making good progress, but I still have a number of suggestions for the tests. These things are tricky.

File: test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/lang/CharacterTest.scala (hunk 20 above)
Comment (er1c) [reply test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/lang/CharacterTest.scala]: I updated this back to a previous version I had that is hopefully more straightforward to understand what is being tested.

File: test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/lang/CharacterTest.scala (hunk 27 above)
Comment (er1c) [reply test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/lang/CharacterTest.scala]: Added this static ranger as `toUpperCase_CodePoint_special_cases`

File: test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/lang/CharacterTest.scala [hunk 28]
```diff
@@ -383,6 +383,352 @@ class CharacterTest {
...
+} yield {
+  //val strUpperCase: String = new String(Array(cp), 0, 1).toUpperCase()
+  s"    assertEquals(${format(chrUpperCase)}, Character.toUpperCase(${format(cp)}))"
+}
+
+val r = new scala.util.Random(100)
+r.shuffle(res).take(100).foreach(println)
+*/
+  @Test def toUpperCase_Int_MultiCodePoints(): Unit = {
+    // 100 "Random" codepoints of the 1Mil+
```
Comment (sjrd) [inline test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/lang/CharacterTest.scala:629]: To be honest, I don't think this is very useful. It would be useful if we implemented the general uppercasing ourselves. But in fact we delegate most of the work to JavaScript's uppercasing.

So what's important to test is:
* Any value that is a special case in our algorithms (:+1: for `toUpperCase_CodePoint_special_cases` and `toUpperCase_CodePoint_StringUpperCase_diff_CharacterUpperCase`)
* A few values to exercise the different code paths in our algorithms. For that I believe the list I gave earlier, with one representative code point in each category, is much more useful than this list of random values. Because that way we exercise each code path. And that should be in `toUpperCase_Int`.

The same comment applies to `toLowerCase_Int`.

Other than that, LGTM (notably, the `toTitleCase` commit with its tests LGTM).

**Review by sjrd — CHANGES_REQUESTED** [review]

**Comment by sjrd** [comment]
When you come back to this PR, would you mind rebasing it on top of the latest master, to get the updated tests for JDK 11? Thanks.

**Comment by er1c** [comment]
@sjrd updated, waiting on CI  results

**Comment by sjrd** [comment]
Thank you. That is excellent.

There's only one rebase problem: the old `toUpperCase_Int_MultiCodePoints` is introduced in the first commit, and removed in the second commit. It should be removed from the first commit.

Otherwise LGTM.

**Comment by er1c** [comment]
Added back the `toUpperCase_Int_MultiCodePoints`

**Comment by sjrd** [comment]
Ach, I meant it should not have been present in *either* commit. Could you please remove it from the first commit, and make sure it doesn't reappear in the second commit.

Sorry for the confusion. I probably wasn't clear enough. I may still not be very clear right now, am I?

**Comment by er1c** [comment]
removed in both :)

**Review by sjrd — APPROVED** [review]
LGTM! Thank you for bearing with me. ;)

**Comment by er1c** [comment]
> LGTM! Thank you for bearing with me. ;)

that's my line!  thank you for your patience with me.

**Comment by er1c** [comment]
pushed newline fix for scalastyle check
